internal/service/alarms/internal/db/migrations/000003_create_alarm_event_record_table.up.sql

```diff
@@ -32,68 +33,102 @@ CREATE TABLE IF NOT EXISTS alarm_event_record (
+        IF NEW.alarm_status = 'resolved' THEN
+            NEW.alarm_changed_time = NEW.alarm_cleared_time;
+            NEW.notification_event_type := 'CLEAR';
+        END IF;
 
-    -- Alarm record cleared. Set alarm_changed_time to alarm_cleared_time
-    IF NEW.perceived_severity = 5 AND NEW.perceived_severity IS DISTINCT FROM OLD.perceived_severity THEN
-        NEW.alarm_changed_time = NEW.alarm_cleared_time;
+        -- alarm_sequence_number is auto-incremented
         RETURN NEW;
-    END IF;
 
-    -- Alarm record acknowledged. Set alarm_changed_time to alarm_acknowledged_time
-    IF NEW.alarm_acknowledged = TRUE AND NEW.alarm_acknowledged IS DISTINCT FROM OLD.alarm_acknowledged THEN
-        NEW.alarm_changed_time = NEW.alarm_acknowledged_time;
-        RETURN NEW;
-    END IF;
+    -- Handle updates to existing alarms
+    ELSIF TG_OP = 'UPDATE' THEN
+        -- 1. Alarm status handling (highest priority)
+        IF NEW.alarm_status = 'resolved' THEN
+            NEW.notification_event_type := 'CLEAR';
+
+            -- Only update change time to cleared time and sequence number if transitioning from a non-resolved state
+            IF OLD.alarm_status IS DISTINCT FROM 'resolved' THEN
+                NEW.alarm_changed_time = NEW.alarm_cleared_time;
+                NEW.alarm_sequence_number := nextval('alarm_sequence_seq');
+             END IF;
 
-    -- Alarm record updated. Set alarm_changed_time to current timestamp
-    IF  NEW.object_id IS DISTINCT FROM OLD.object_id OR
-        NEW.object_type_id IS DISTINCT FROM OLD.object_type_id OR
-        NEW.alarm_definition_id IS DISTINCT FROM OLD.alarm_definition_id OR
-        NEW.probable_cause_id IS DISTINCT FROM OLD.probable_cause_id
-    THEN
-        NEW.alarm_changed_time = CURRENT_TIMESTAMP;
+        -- 2. Handling alarm_acknowledged. Set alarm_changed_time to alarm_acknowledged_time
+        ELSIF NEW.alarm_acknowledged THEN
+            NEW.notification_event_type := 'ACKNOWLEDGE';
+
+             -- Update sequence only on first acknowledgment
+            IF NEW.alarm_acknowledged IS DISTINCT FROM OLD.alarm_acknowledged THEN
+                NEW.alarm_changed_time = NEW.alarm_acknowledged_time;
+                NEW.alarm_sequence_number := nextval('alarm_sequence_seq');
+            END IF;
+
+        -- 3. Other changes (only if not acknowledged)
+        ELSIF NOT NEW.alarm_acknowledged THEN
+            IF (NEW.object_id IS DISTINCT FROM OLD.object_id OR
+                NEW.object_type_id IS DISTINCT FROM OLD.object_type_id OR
+                NEW.alarm_definition_id IS DISTINCT FROM OLD.alarm_definition_id OR
+                NEW.probable_cause_id IS DISTINCT FROM OLD.probable_cause_id)
+            THEN
+                NEW.notification_event_type := 'CHANGE';
+                NEW.alarm_changed_time := CURRENT_TIMESTAMP;
+                NEW.alarm_sequence_number := nextval('alarm_sequence_seq');
+            END IF;
+        END IF;
+
+        RETURN NEW;
```

Comment:
We've discussed this briefly in the past, but I feel like it is worth mentioning again ... I'm a bit concerned that we're using the AlarmEventRecord table to do double-duty in also storing event/notification type information.   If two or more changes occur to any given alarm in a short period of time then we may not be able to send off both notifications and the client would only get the last event.   Clients may want to see all transitions as they may want to take action on each transition.

Reply:
(just noting it here what we discussed) 

we do track all the changes back to back...partially. e.g if we have a subscriber that has a filter "CHANGE" (so notify me for everything that's NOT change), and then we have a notification that was firing, and the resolved and then firing again and now the subscriber is added, the subscriber will receive a payload with two events - one for the resolved and one for firing. The first one resolved (updated from firing to resolved) and then finally the second (firing with the new start time but same fingerprint, resulting in a new row/event) 

But its true 3 things happened but the client is getting 2. But given that its currently resolved..it somewhat implied that originally it was firing?